### ISSUE TITLE
syzkaller: memory leak in `subflow_create_ctx`

### ISSUE THREAD

**Issue by cpaasch** [multipath-tcp/mptcp_net-next#356]
Head: 89ce2ca8c923 (02/28)
30384974730d28e (02/27)
0150d5be08985f6a

Trace:
```
BUG: memory leak
unreferenced object 0xffff88812df5a700 (size 256):
  comm "syz-executor.3", pid 2701, jiffies 4297168679 (age 63.966s)
  hex dump (first 32 bytes):
    00 01 00 00 00 00 ad de 22 01 00 00 00 00 ad de  ........".......
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  backtrace:
    [<00000000109c3810>] kmalloc_trace+0x21/0x90 mm/slab_common.c:1062
    [<00000000a01124a9>] kmalloc include/linux/slab.h:580 [inline]
    [<00000000a01124a9>] kzalloc include/linux/slab.h:720 [inline]
    [<00000000a01124a9>] subflow_create_ctx+0x5b/0xf0 net/mptcp/subflow.c:1752
    [<00000000a180cdc9>] subflow_ulp_init+0x3a/0x140 net/mptcp/subflow.c:1832
    [<00000000cf1a8bf5>] __tcp_set_ulp net/ipv4/tcp_ulp.c:146 [inline]
    [<00000000cf1a8bf5>] tcp_set_ulp+0x142/0x230 net/ipv4/tcp_ulp.c:167
    [<00000000382f383a>] mptcp_subflow_create_socket+0x171/0x300 net/mptcp/subflow.c:1715
    [<000000002fe382d9>] __mptcp_socket_create net/mptcp/protocol.c:89 [inline]
    [<000000002fe382d9>] __mptcp_nmpc_socket.part.0+0x83/0x1e0 net/mptcp/protocol.c:122
    [<00000000fa3a0e61>] __mptcp_nmpc_socket net/mptcp/protocol.c:115 [inline]
    [<00000000fa3a0e61>] mptcp_bind+0x55/0xe0 net/mptcp/protocol.c:3714
    [<00000000e3b192be>] __sys_bind+0x89/0x100 net/socket.c:1801
    [<00000000f1ffd90b>] __do_sys_bind net/socket.c:1812 [inline]
    [<00000000f1ffd90b>] __se_sys_bind net/socket.c:1810 [inline]
    [<00000000f1ffd90b>] __x64_sys_bind+0x18/0x20 net/socket.c:1810
    [<000000000703658a>] do_syscall_x64 arch/x86/entry/common.c:50 [inline]
    [<000000000703658a>] do_syscall_64+0x37/0x90 arch/x86/entry/common.c:80
    [<00000000d12f04a9>] entry_SYSCALL_64_after_hwframe+0x72/0xdc
```

No reproducers yet.

Kconfig: [Kconfig_k8_kmemleak.txt](https://github.com/multipath-tcp/mptcp_net-next/files/10736903/Kconfig_k8_kmemleak.txt)

**Comment by matttbe** [comment]
@cpaasch this issue has been fixed we think but hard to confirm without reproducer.

**Comment by matttbe** [comment]
FYI, from IRC, Paolo pointed out that this issue was caused by "mptcp: drop legacy code", requiring "mptcp: fix possible memory leak on syn_recv race" to fix this issue but then causing #370.

It has now been reverted:

- 4ef602b23ce51: Revert "mptcp: drop legacy code"
